src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -372,14 +394,20 @@ def process_messages(
                 *tags.keys(),
                 *tags.values(),
             }
-            org_strings[org_id].update(parsed_strings)
+            non_org_zero = parsed_strings.difference(SHARED_STRINGS)
+            org_zero = parsed_strings.intersection(SHARED_STRINGS)
+
+            org_strings[org_id].update(non_org_zero)
+            org_strings[0].update(org_zero)
```

Comment:
Would it make sense to create a documented constant somewhere? For example

```python
#: org_id used for strings that are shared across organizations
ORG_ID_SHARED = 0
```

Reply:
Another thought: Would it make sense to make `org_id=0` an implementation detail of the indexer? This would mean you could simply write `org_strings[org_id].update(parsed_strings)` again. The downside would be that the return value of `bulk_record` would contain a lot of duplicates (for example `{1: {"release": 789}, 2: {"release": 789}}`. But it would simplify the code in `multiprocess.py`, because you wouldn't have to deal with `org_id=0`/`SHARED_STRINGS` at all.

Reply:
@jjbayer I think that it could simplify `multiprocess.py` but it will make `bulk_record` more complicated. However, I currently have the `resolve` method handle changing the org_id to `0` so I dk if that is confusing. I don't necessarily think that the indexer needs to handle that logic but maybe it should have a check before creating new values that if the string is in the set of shared strings that the org is `0`.

Reply:
Yeah the `resolve` method is what got me thinking.. when we call `record` or `bulk_record` from any other place than `multiprocess.py` (e.g. tests), the caller has to perform the same dance as `multiprocess.py`. You would expect `indexer.record(org_id, "release")` and `indexer.resolve(org_id, "release")` to return the same integer, but they now return different values because one maps `org_id` to `0` and one does not. Although the assertion you introduced also prevents that, so I guess we're good.

Reply:
+1 to what @jjbayer said. I don't think this logic should be in the consumer at all.

My arguments to move all this logic within the indexer are:

- This way you are actually sending the org id 0 string through the main pipeline, which means caching, read from DB and write to DB. While this is likely a minor performance concern if at all, we will have these values in the DB and then we will have to deal with the TTL. I think we should not even have them going into the DB.
- The idea of shared strings that are not organization specific is an implementation detail of the postgres implementation. If you put them in the consumer you are assuming that any implementation of the indexer will have to deal with shared strings, this is a constraint we do not need. If we push them into the indexer implementation you reduce the impact of this feature.
- You are now mixing the code that manages different types of strings with the code that manages the overall flow in the consumer, thus this code does not have a single responsibility anymore.

Ideally the consumer could live its life without caring about where the strings come from. In practice we will still have to surface from the indexer this detail for the message to be reassembled (see Oliver's proposal) but that would minimally impact the code of the consumer.

> it will make bulk_record more complicated.

Not necessarily. There is a way to clean that up in an elegant way. Make the indexer implementaiton a delegate: https://en.wikipedia.org/wiki/Delegation_pattern.  You wrap the indexer into another one that only deals with the static strings and then delegate to the postgres indexer the resolution/record/etc of all the others. The current indexer does not get changed, Just the external level does.